contracts/ProductNft/ProductNft.sol

```diff
@@ -0,0 +1,170 @@
+*/
+
+contract ProductNft is ERC1155 {
+    using Counters for Counters.Counter;
+
+    string public baseURI;
+    uint16 totalAmountOfTokenIds;
+    uint16 rarestTokensLeft;
+    uint16 rarerTokensLeft;
+    uint16 rareTokensLeft;
+    uint8 constant rarityRarest = 1;
+    uint8 constant rarityRarer = 2;
+    uint8 constant rarityRare = 3;
+    Counters.Counter public rarestTokenIds;
+    Counters.Counter public rarerTokenIds;
+    Counters.Counter public rareTokenIds;
+    uint256 public constant PRICE_PER_RAREST_TOKEN = 1.5 ether;
+    uint256 public constant PRICE_PER_RARER_TOKEN = 0.55 ether;
+    uint256 public constant PRICE_PER_RARE_TOKEN = 0.2 ether;
+    uint16 public startRarerTokenIdIndex;
+    uint16 public startRareTokenIdIndex;
+
+  mapping(uint => string) public _URIS;
+
+  constructor( 
+    string memory baseURI_,   
+    uint16 _startRarerTokenIdIndex,
+    uint16 _startRareTokenIdIndex,
+    uint16 _totalAmountOfTokenIds
+    ) ERC1155(baseURI_) {
+        baseURI = baseURI_;
+        startRarerTokenIdIndex = _startRarerTokenIdIndex;
+        startRareTokenIdIndex = _startRareTokenIdIndex;
+        totalAmountOfTokenIds = _totalAmountOfTokenIds;
+        rarestTokensLeft = _startRarerTokenIdIndex;
+        rarerTokensLeft = _startRareTokenIdIndex - _startRarerTokenIdIndex; 
+        rareTokensLeft = totalAmountOfTokenIds - startRareTokenIdIndex;
+    }
+
+    /**
+    *@dev   This function allows the generation of a URI for a specific token Id: baseURI/tokenId.json 
+    *       if it does not exist already. If it does exist, that token URI will be returned.
+    *@param _tokenId is the token Id to generate or return the URI for.     
+    */
+    function _URI(uint256 _tokenId) public view returns (string memory) {
+      if(bytes(_URIS[_tokenId]).length != 0) {
+        return string(_URIS[_tokenId]);
+      }
+      return string(abi.encodePacked(baseURI, Strings.toString(_tokenId), ".json"));
+    }
+    
+    /**
+    *@dev   Using this function, a given amount will be turned into an array.
+    *       This array will be used in ERC1155's batch mint function. 
+    *@param amount is the amount provided that will be turned into an array.
+    */
+    function _turnAmountIntoArray(uint16 amount) internal pure returns (uint256[] memory tokenAmounts) {
+        tokenAmounts = new uint[](amount);
+        for (uint256 i = 0; i < amount;) {
+            tokenAmounts[i] = i + 1;
+            unchecked {
+                i++;
+            }
+        }
+    }
+
+    /**
+    *@dev   A counter is used to track the token Ids that have been minted.
+    *       A token Id will be returned and turned into an array. 
+    *@param rarity The rarity defines at what point the count starts.
+    *       Using requirements, rarities have been assigned a set of tokenIds. 
+    */
+    function _countBasedOnRarity(uint8 rarity) internal returns (uint256 tokenId) {
```

Comment:
I don't know if it would be an issue here, but maybe let's do a check that it is one of the valid values. 
I would actually use an enum instead of a uint8 for the rarity calculations.